src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/matcher/SymbolicRegexMatcher.cs

```diff
@@ -469,24 +454,17 @@ private State<S> CreateNewTransition(State<S> q, S atom, int offset)
             lock (this)
             {
                 // check if meanwhile delta[offset] has become defined possibly by another thread
-                State<S> p = delta[offset];
+                State<S> p = builder.delta[offset];
                 if (p != null)
                     return p;
                 else
                 {
-                    // this is the only place in code where the Next method is called
-                    p = q.Next(atom);
-                    // if the statearray was extended then delta must be extended accordingly
-#if DEBUG
-                    if (StateLimit > builder.statearray.Length)
-                        throw new AutomataException(AutomataExceptionKind.InternalError_SymbolicRegex);
-#endif
-                    if (StateLimit < builder.statearray.Length)
-                    {
-                        StateLimit = builder.statearray.Length;
-                        Array.Resize(ref delta, StateLimit << K);
-                    }
-                    delta[offset] = p;
+                    // this is the only place in code where the Next method is called in the matcher
+                    p = q.Next(atom, antimirov);
+                    builder.delta[offset] = p;
+                    //switch to antimirov mode if the maximum bound has been reached
+                    if (p.Id == s_STATEMAXBOUND)
+                        antimirov = true;
```

Comment:
Regarding getting the antimirov check out of the inner loop, throwing an exception here and catching it outside the matching loop to call into an antimirov specific version might be a way to avoid complicating the control flow.

Reply:
See prior comment, I will move the antimirov flag from the matcher to the builder. In this way it is never passed around as an argument and only used in the derivative code when **OR** is involved, because the builder is shared in all of this. After I do this commit, you could rerun one of the experiments to confirm that there are no perf regressions.

Reply:
Sounds good!